src/Tile/Syntax/Lifted.hs

```diff
@@ -0,0 +1,166 @@
+  => (m :.: i) expr ::: (m :.: i) expr
+  -> (forall j . Permutable j => (i :.: j) expr -> (m :.: i :.: j) expr)
+  -> (m :.: i) expr
+let' v f = let'' v (f . C . pure)
+
+let''
+  :: (Applicative m, S.Let expr, Permutable i)
+  => (m :.: i) expr ::: (m :.: i) expr
+  -> (forall j . Permutable j => j expr -> (m :.: i :.: j) expr)
+  -> (m :.: i) expr
+let'' (tm ::: ty) f = S.let' <$> ((:::) <$> tm <*> ty) <*> mapC (fmap getC) (f id)
+
+
+-- Lam
+
+lam :: (Applicative m, S.Lam expr, Permutable i) => (m :.: i) Plicit -> (forall j . Permutable j => (i :.: j) expr -> (m :.: i :.: j) expr) -> (m :.: i) expr
+lam p f = S.lam <$> p <*> mapC (fmap getC) (f (C (pure id)))
+
+($$) :: (Applicative m, S.Lam expr) => m expr -> m expr -> m expr
+($$) = liftA2 (S.$$)
+
+infixl 9 $$
+
+
+-- Type
+
+type' :: (Applicative m, S.Type expr) => m expr
+type' = pure S.type'
+
+(>->) :: (Applicative m, S.Type expr, Permutable i) => ((m :.: i) Plicit, (m :.: i) expr) -> (forall j . Permutable j => (i :.: j) expr -> (m :.: i :.: j) expr) -> (m :.: i) expr
+(pl, a) >-> b = (S.>->) <$> ((,) <$> pl <*> a) <*> mapC (fmap getC) (b (C (pure id)))
+
+infixr 6 >->
+
+(-->) :: (Applicative m, S.Type expr, Permutable i) => (m :.: i) expr -> (m :.: i) expr -> (m :.: i) expr
+a --> b = (pure Ex, a) >-> const (weakens b)
+
+infixr 6 -->
+
+(==>) :: (Applicative m, S.Type expr, Permutable i) => (m :.: i) expr -> (forall j . Permutable j => (i :.: j) expr -> (m :.: i :.: j) expr) -> (m :.: i) expr
+a ==> b = (pure Im, a) >-> b
+
+infixr 6 ==>
+
+
+-- Prob
+
+ex :: (Applicative m, S.Prob expr, Permutable i) => (m :.: i) expr -> (forall j . Permutable j => (i :.: j) expr -> (m :.: i :.: j) expr) -> (m :.: i) expr
+ex t f = S.ex <$> t <*> mapC (fmap getC) (f (C (pure id)))
+
+(===) :: (Applicative m, S.Prob expr) => m expr ::: m expr -> m expr ::: m expr -> m expr
+(tm1 ::: ty1) === (tm2 ::: ty2) = (S.===) <$> ((:::) <$> tm1 <*> ty1) <*> ((:::) <$> tm2 <*> ty2)
+
+infixl 4 ===
+
+
+-- Elaborator scripts
+
+runScript :: Functor m => Script a m a -> m a
+runScript = strengthen . (`getScript` id)
+
+runCScript :: Functor m => (Script (i a) m :.: i) a -> (m :.: i) a
+runCScript = mapC runScript
+
+throw
+  :: (Applicative m, Applicative hw)
+  => (forall h . Permutable h => ((m :.: hw) :.: h) a -> ((m :.: hw) :.: h) w)
+  -> m a
+  -> (m :.: hw) w
+throw k = strengthen . k . liftC . liftC
+
+reset :: Applicative m => Script a m a -> Script w m a
+reset m = Script $ \ k -> throw k $ runScript m
+
+resetC :: Applicative m => (Script (i a) m :.: i) a -> (Script w m :.: i) a
+resetC = mapC reset
+
+liftScript :: Functor m => m a -> Script w m a
+liftScript m = Script $ \ k -> strengthen (k (liftC (liftC m)))
+
+liftCScript :: Functor m => (m :.: i) a -> (Script w m :.: i) a
+liftCScript = mapC liftScript
+
+newtype Script t m a = Script
+  { getScript
+    :: forall hw
+    .  Permutable hw
+    => (forall h . Permutable h => ((m :.: hw) :.: h) a -> ((m :.: hw) :.: h) t)
+    -> (m :.: hw) t
+  }
```

Comment:
This is the `CPSA` type from the paper. I couldn’t figure out how to make the clearer definition work using the transitivity of `Extends`, which they alluded to being possible but never published.

I haven’t really used this yet since scope safety precludes a `Monad` instance (and thus `do`-notation, in the absence of `-XRebindableSyntax`), which is the entire reason why I use `Script` thus far.

That said, in the paper they get a lot of mileage out of this for scope-safe effectful transformations (e.g. moving code across binders), so I’m leaving it in here so that I can use it in e.g. the solver.